### PR TITLE
storage/inmem: allow opt out of object roundtrip

### LINKED ISSUE #4708: Allow for adding objects to storage without round-tripping through JSON
## What part of OPA would you like to see improved?
Motivating issue: https://github.com/open-policy-agent/gatekeeper/issues/2060

When adding data to the in-memory OPA storage object, `inmem.store.Write` round-trips the incoming object through JSON to ensure the object may be validly added to OPA storage. Adding an object which cannot be serialized to JSON could cause all sorts of misbehaviors when running OPA policies.

From `inmem.store.Write`:
```
func (db *store) Write(_ context.Context, txn storage.Transaction, op storage.PatchOp, path storage.Path, value interface{}) error {
	underlying, err := db.underlying(txn)
	if err != nil {
		return err
	}
	val := util.Reference(value)
	if err := util.RoundTrip(val); err != nil { // <-- this function
		return err
	}
	return underlying.Write(op, path, *val)
}
```

Unfortunately, this round-tripping is very slow compared to the rest of what needs to be done to store/update the new data.

Benchmark with round-tripping:
```
~/frameworks/constraint$ go test ./pkg/client/drivers/local/ --run=NONE --bench=. --benchtime=10s --cpuprofile=cpu.out
goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkStorages_AddData-16    	   29972	    334523 ns/op
PASS
ok  	github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local	17.951s
```

Benchmark without round-tripping (with the check commented out):
```
~/frameworks/constraint$ go test ./pkg/client/drivers/local/ --run=NONE --bench=. --benchtime=10s --cpuprofile=cpu.out
goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local
cpu: Intel(R) Xeon(R) CPU @ 2.20GHz
BenchmarkStorages_AddData-16    	 4121306	      2765 ns/op
PASS
ok  	github.com/open-policy-agent/frameworks/constraint/pkg/client/drivers/local	15.388s
```

image

In the benchmark >99% of the time spend adding new objects is in round-tripping through JSON.

## Describe the ideal solution
Ideally, there would be some way for callers to add data to `inmem.store` without needing to do this round-tripping. Callers may be able to do this validation themselves more cheaply, or are operating in an environment where all objects by definition are serializable to JSON (e.g. the incoming request was JSON, and that is the object to be put into OPA storage) and so this validation is unnecessary.

This option could take many forms:

- An option on `inmem.store` at creation time to not round-trip objects through JSON
- An option on `Write()` to skip this check
- A separate function `WriteNoRoundTrip()` which is the same as `Write()` but does not include the check

## Additional Context
Per the motivating bug - the issue which (at least) one user of Gatekeeper is having is that the speed of these writes means that it takes a long time for queries to run since it takes too long to acquire a read lock for the Storage object (which is required for Query to run).

### PR DESCRIPTION
**Pull request by srenatus**
This is a continuance of @willbeason's work in https://github.com/open-policy-agent/opa/pull/4709. 

It will allow everyone to use an in-memory store _without roundtripping_. Before, we had been roundtripping every object written to that store through json, to canonicalize its data type. However, for many code paths, this isn't necessary. Notably, when our own http handlers take in JSON through encoding/json and unmarshal it, it'll always be in the right format, and the extra roundtrip can be avoided.

This roundtrip would also create a deep copy of the object, and with the opt-out, that no longer happens.

Fixes #4708.

### CODE REVIEW

**Comment by srenatus** [comment]
@ashutosh-narkar I've requested your review because I had changed some things in the bundle plugin tests. I think we're clear -- whenever a bundle is read, it'll go through encoding/json, and thus only have "safe" types, like, using `[]interface{}{"string1", "string2"}` and not `[]string{"string1", "string2"}`. It's only the test setup that necessitated the changes.

File: internal/presentation/presentation_test.go
Comment (srenatus) [inline internal/presentation/presentation_test.go:22]: I think I've flicked all (or most) tests to use the non-roundtripping inmem store. It just goes through smoothly, so we're probably OK wrt how it works in OPA-as-server.

File: loader/loader.go
Comment (srenatus) [inline loader/loader.go:57]: I couldn't find this one being used in our code bases, but I figured for completeness' sake, we could still add 👇

File: plugins/bundle/plugin_test.go
Comment (srenatus) [inline plugins/bundle/plugin_test.go:406]: Before, this would be roundtripped though json, and it would be _turned into_ `[]interface{}` (see [playground](https://go.dev/play/p/pqyodPVAQUa)). Now, we're feeding in the proper value types.

In real life, this shouldn't matter, since when bundles are read, their `Value`s should come from a json string unmarshalled using encoding/json, so we'd only ever get `[]interface{}`, never `[]string`, there.

File: plugins/bundle/plugin_test.go
Comment (philipaconrad) [inline plugins/bundle/plugin_test.go:452]: This was a small, but good fix. :+1: 

The `err` check is for the `Store` a few lines up, and `MustUnmarshalJSON` doesn't return an error, so an independent check for its error condition (wrong result) is put in a few lines down. This is clearer than the "chained-if" style that was there before, I think.

Comment (srenatus) [reply plugins/bundle/plugin_test.go:452]: I think that's something a linter should be able to catch, anyways. If the if's branch breaks the control flow (like, `return`, or `t.Fatal` or `panic`), then there's no need to wrap the other conditionals into `else`. But maybe it's just a pet peeve of mine.

File: storage/inmem/opts.go
Comment (philipaconrad) [inline storage/inmem/opts.go:18]:
```suggestion
// Callers should disable this if they cannot guarantee all objects passed to
// Write() are serializable to JSON. Failing to do so may result in undefined
// behavior, including panics.
//
// If setting to false, callers should consider deep-copying any objects passed
// to Write, unless they can guarantee the objects will not be mutated after
// being written to the store.
```

File: storage/inmem/test/testutil.go
Comment (philipaconrad) [inline storage/inmem/test/testutil.go:18]:
```suggestion
// NewFromObject returns an inmem store from the passed object, with some
```
Just one space extra space character lurking there. :smile:

File: storage/inmem/opts.go
Comment (srenatus) [reply storage/inmem/opts.go:18]: I've updated the comment in a later commit -- what do you think about that one?

File: plugins/bundle/plugin_test.go
Comment (philipaconrad) [reply plugins/bundle/plugin_test.go:406]: The Go Playground example you included nicely illustrates your point. This change makes sense.

File: storage/inmem/opts.go
Comment (philipaconrad) [reply storage/inmem/opts.go:18]: One second... I'll go see the latest revisions. :smile:

Comment (philipaconrad) [reply storage/inmem/opts.go:18]: I'm a little confused by the "they" on line 23. I assume it's referring to the caller, and that they have a responsibility to ensure no mutations are happening to objects going into the store?

Comment (srenatus) [reply storage/inmem/opts.go]: They need to ensure that if the object are mutated (in the store) it won't matter for them. Basically "hand them over"... I'm grateful for all suggestions here 😃

**Comment by srenatus** [comment]
Ideally, the code examples in the inmem/example_test.go would be updated, too. But I didn't get to that, and didn't want to push this further 😅

File: storage/inmem/opts.go
Comment (philipaconrad) [reply storage/inmem/opts.go]: I added some extra words in the suggested change block to make it clear it's the caller we're talking about, because when I hear "consider \<action\>" by itself, I usually parse that as a command directed at me, instead of another target.

I know I'm bikeshedding a bit there. :bike: It's probably fine, but clarity is always preferred in comments that may be there for a bit. :smile:

**Comment by philipaconrad** [comment]
Looks like DCO blew up because of a change in email identity of the committer. :slightly_frowning_face:

@srenatus I think your GMail and Styra emails got mixed somewhere in the commit history, and confused the bot.

**Comment by philipaconrad** [comment]
> Ideally, the code examples in the inmem/example_test.go would be updated, too. But I didn't get to that, and didn't want to push this further sweat_smile

Understandable. I don't see anything worrying in these commits, but definitely ping @ashutosh-narkar again if they haven't had time to see these changes.

I'm good to mark Approve in a bit once tests go green. :+1:

File: storage/inmem/opts.go
Comment (philipaconrad) [inline storage/inmem/opts.go:18]:
```suggestion
// If setting to false, callers should consider deep-copying any objects passed to
// Write() unless they can guarantee the objects will not be mutated after being written.
```

**Review by philipaconrad — APPROVED** [review]
I've only got [nit]-level suggestions on this PR. The rest looks great! :smile:

Marking Approve, so that @srenatus can get it merged when they have time to look at it tomorrow.

**Review by ashutosh-narkar — DISMISSED** [review]

**Comment by srenatus** [comment]
I'll update the comment and merge this!

**Pull request merged**
